File: Haskell/TCPServer.hs
--If anyone would like to fix this code (if that is even possible) please for the love of all things good and holy on this earth do so.
module Main where
 import Data.Bits
 import Foreign.C
 import Foreign.C.Types
 import Foreign.Ptr
 import Foreign.Storable
 import Foreign.Marshal.Alloc
 foreign import ccall "socket"     c_socket::CInt->CInt->CInt->IO CInt
 foreign import ccall "setsockopt" c_setsockopt::CInt->CInt->CInt->Ptr CInt->CInt->IO CInt
 foreign import ccall "bind"       c_bind::CInt->Ptr a->CInt->IO CInt
 foreign import ccall "listen"     c_listen::CInt->CInt->IO CInt
 foreign import ccall "accept"     c_accept::CInt->Ptr a->Ptr CInt->IO CInt
 foreign import ccall "read"       c_read::CInt->Ptr a->CInt->IO CInt
 foreign import ccall "write"      c_write::CInt->Ptr a->CInt->IO CInt
 foreign import ccall "close"      c_close::CInt->IO CInt
 main=do
  true<-mallocBytes 4 -- Allocating new memory because ??? HOW IN CHRIST DO I MAKE A STRUCT IN A SENSIBLE WAY?
  addrsize<-mallocBytes 4
  addr<-mallocBytes 16
  buffer<-mallocBytes 256
  poke (true::Ptr CInt) 1
  poke (addr::Ptr CShort) 2
  poke ((plusPtr addr 2)::Ptr CUShort) (shiftR 3721 8) -- Have to CUShort because for SOME GOD FORSAKEN REASON IT GETS CASTED TO CHAR BEFORE BEING SHIFTED?
  poke ((plusPtr addr 3)::Ptr CUChar) ((.&.) 3721 0xFF)
  server<-c_socket 2 1 0
  result<-c_setsockopt server 1 15 true 4
  result<-c_bind server addr 16
  result<-c_listen server 5
  mainloop server addr addrsize buffer
 mainloop server addr addrsize buffer=do -- IS THIS GOING TO STACK OVERFLOW? IS THERE NO SENSIBLE WAY TO MAKE A LOOP???
  client<-c_accept server addr addrsize
  size<-c_read client buffer 0xFF
  result<-c_write 1 buffer size
  please <- newCString "Message recieved\n"
  result<-c_write client please 17
  result<-c_close client
  mainloop server addr addrsize buffer

File: Fortran/TCPServer.f90
program tcpserver
 use iso_c_binding !,only:c_int8_t,c_short,c_int,c_long,c_ptr,c_loc
 implicit none
 interface
  integer(c_int) function socket(family,type,protocol) bind(c,name='socket')
   import::c_int
   integer(c_int),value,intent(in)::family,type,protocol
   end function socket
  integer(c_int) function setsockopt(socket,layer,option,value,size) bind(c,name='setsockopt')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::socket,layer,option,size
   type(c_ptr),value,intent(in)::value
   end function setsockopt
  integer(c_int) function bind(socket,address,length) bind(c,name='bind')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::socket,length
   type(c_ptr),value,intent(in)::address
   end function bind
  integer(c_int) function listen(socket,backlog) bind(c,name='listen')
   import::c_int
   integer(c_int),value,intent(in)::socket,backlog
   end function listen
  integer(c_int) function accept(socket,address,addrsize) bind(c,name='accept')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::socket
   type(c_ptr),value,intent(in)::address,addrsize
   end function accept
  integer(c_int) function read(fd,buffer,size) bind(c,name='read')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::fd,size
   type(c_ptr),value,intent(in)::buffer
   end function read
  integer(c_int) function write(fd,buffer,size) bind(c,name='write')
   import::c_int,c_ptr
   integer(c_int),value,intent(in)::fd,size
   type(c_ptr),value,intent(in)::buffer
   end function write
  integer(c_int) function close(fd) bind(c,name='close')
   import::c_int
   integer(c_int),value,intent(in)::fd
   end function close
  end interface
 integer(c_int8_t),dimension(256),target::buffer
 character(len=17),target::reply="Message received"//new_line('1') ! no clue what the parameter does for newline()
 type::sockaddr
  integer(c_short)::type=2
  integer(c_int8_t)::porthi=SHIFTR(3721,8)
  integer(c_int8_t)::portlo=IAND(3721,255)
  integer(c_int8_t),dimension(4)::ip=[0,0,0,0]
  integer(c_int8_t),dimension(8)::pad
  end type sockaddr
 type(sockaddr),target::address
 integer(c_int)::server,client,result,size
 integer(c_int),target::true=1,addrsize=16
 server=socket(2,1,0)
 result=setsockopt(server,1,15,c_loc(true),4)
 result=bind(server,c_loc(address),addrsize)
 result=listen(server, 5)
 do while(1==1)
  client=accept(server,c_loc(address),c_loc(addrsize))
  size=read(client,c_loc(buffer),255)
  size=write(1,c_loc(buffer),size)
  size=write(client,c_loc(reply),17)
  result=close(client)
  end do
 end program tcpserver